cli/lsp/diagnostics.rs

```diff
@@ -296,57 +297,70 @@ struct ChannelUpdateMessage {
   batch_index: Option<usize>,
 }
 
-#[derive(Clone, Debug)]
+#[derive(Debug)]
 struct SpecifierState {
-  has_no_cache_diagnostic: bool,
+  version: Option<i32>,
+  no_cache_diagnostics: Vec<lsp::Diagnostic>,
 }
 
-#[derive(Clone, Debug, Default)]
+#[derive(Debug, Default)]
 pub struct DiagnosticsState {
-  specifiers: HashMap<ModuleSpecifier, (Option<i32>, SpecifierState)>,
+  specifiers: RwLock<HashMap<ModuleSpecifier, SpecifierState>>,
 }
 
 impl DiagnosticsState {
   fn update(
-    &mut self,
+    &self,
     specifier: &ModuleSpecifier,
     version: Option<i32>,
     diagnostics: &[lsp::Diagnostic],
   ) {
-    let current_version = self.specifiers.get(specifier).and_then(|s| s.0);
-    let is_new = match (version, current_version) {
-      (Some(arg), Some(existing)) => arg >= existing,
-      _ => true,
-    };
-    if is_new {
-      self.specifiers.insert(
-        specifier.clone(),
-        (
-          version,
-          SpecifierState {
-            has_no_cache_diagnostic: diagnostics.iter().any(|d| {
-              d.code
-                == Some(lsp::NumberOrString::String("no-cache".to_string()))
-                || d.code
-                  == Some(lsp::NumberOrString::String(
-                    "no-cache-npm".to_string(),
-                  ))
-            }),
-          },
-        ),
-      );
+    let mut specifiers = self.specifiers.write();
+    let current_version = specifiers.get(specifier).and_then(|s| s.version);
+    match (version, current_version) {
+      (Some(arg), Some(existing)) if arg < existing => return,
+      _ => {}
     }
+    let mut no_cache_diagnostics = vec![];
+    for diagnostic in diagnostics {
+      if diagnostic.code
+        == Some(lsp::NumberOrString::String("no-cache".to_string()))
+        || diagnostic.code
+          == Some(lsp::NumberOrString::String("no-cache-npm".to_string()))
+      {
+        no_cache_diagnostics.push(diagnostic.clone());
+      }
+    }
+    specifiers.insert(
+      specifier.clone(),
+      SpecifierState {
+        version,
+        no_cache_diagnostics,
+      },
+    );
   }
 
-  pub fn clear(&mut self, specifier: &ModuleSpecifier) {
-    self.specifiers.remove(specifier);
+  pub fn clear(&self, specifier: &ModuleSpecifier) {
+    self.specifiers.write().remove(specifier);
   }
 
-  pub fn has_no_cache_diagnostic(&self, specifier: &ModuleSpecifier) -> bool {
+  pub fn has_no_cache_diagnostics(&self, specifier: &ModuleSpecifier) -> bool {
     self
       .specifiers
+      .read()
       .get(specifier)
-      .map_or(false, |s| s.1.has_no_cache_diagnostic)
+      .map_or(false, |s| !s.no_cache_diagnostics.is_empty())
```

Comment:
Just a side note: `map_or`s are hard to understand because the arguments are backwards. I previously opened https://github.com/denoland/deno/pull/18212